--- rpm/generic/zfs-kmod.spec.in	2023-11-30 23:48:38.955131592 +0000
+++ ../zfs-kmod.spec.in-sign	2024-01-06 19:14:10.619938449 +0000
@@ -150,6 +150,30 @@
 done
 
 
+# Module signing (modsign)
+#
+# This must be run _after_ find-debuginfo.sh runs, otherwise that will strip
+# the signature off of the modules.
+# (Based on Fedora's kernel.spec workaround)
+%define __modsign_install_post \
+  sign_pem="%{ksrc}certs/signing_key.pem"; \
+  sign_x509="%{ksrc}certs/signing_key.x509"; \
+  if [ -f "${sign_x509}" ]\
+  then \
+    echo "Signing kernel modules ..."; \
+    for kmod in $(find ${RPM_BUILD_ROOT}%{kmodinstdir_prefix}/*/extra/zfs/ -name \*.ko); do \
+      %{ksrc}scripts/sign-file sha256 ${sign_pem} ${sign_x509} ${kmod}; \
+    done \
+  fi \
+%{nil}
+
+# hack to ensure singing happens after find-debuginfo.sh runs
+%define __spec_install_post \
+  %{?__debug_package:%{__debug_install_post}}\
+  %{__arch_install_post}\
+  %{__os_install_post}\
+  %{__modsign_install_post}
+
 %install
 rm -rf ${RPM_BUILD_ROOT}
 
